### PR TITLE
[ENH] Dobin basis transformation

### PR DESCRIPTION
**Pull request by KatieBuc**
Implementation of Issue #3371

Distance based Outlier BasIs using Neighbors (DOBIN) is a new approach to select a set of basis vectors tailored for outlier detection. The transformation outputs are the new coordinate system (such that outliers are best retained in lower the dimensional feature space).

FYI @Lovkush-A @fkiraly @lmmentel

### CODE REVIEW

**Comment by KatieBuc** [comment]
@fkiraly @aiwalter @Lovkush-A @lmmentel this is ready for review :)

File: sktime/transformations/series/dobin.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    DOBIN is a pre-processing algorithm that constructs a set of basis
+    basistors tailored for outlier detection as described by _[1]. DOBIN
+    has a simple mathematical foundation and can be used as a dimension
+    reduction tool for outlier detection tasks.
+
+    Method assumes normalized data, for example:
+    from sklearn.preprocessing import MinMaxScaler, RobustScaler, StandardScaler
+
+    This prevents variables with large variances having disproportional
+    influence on Euclidean distances. The original implelemtation -[1] uses
```
Comment (lmmentel) [inline sktime/transformations/series/dobin.py:31]:
```suggestion
    influence on Euclidean distances. The original implelemtation _[1] uses
```

File: sktime/transformations/series/dobin.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+class DOBIN(BaseTransformer):
+    """Distance based Outlier BasIs using Neighbors (DOBIN).
+
+    DOBIN is a pre-processing algorithm that constructs a set of basis
+    basistors tailored for outlier detection as described by _[1]. DOBIN
+    has a simple mathematical foundation and can be used as a dimension
+    reduction tool for outlier detection tasks.
+
+    Method assumes normalized data, for example:
+    from sklearn.preprocessing import MinMaxScaler, RobustScaler, StandardScaler
```
Comment (lmmentel) [inline sktime/transformations/series/dobin.py:28]:
```suggestion
    ``from sklearn.preprocessing import MinMaxScaler, RobustScaler, StandardScaler``
```

File: sktime/transformations/series/dobin.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+    This prevents variables with large variances having disproportional
+    influence on Euclidean distances. The original implelemtation -[1] uses
+    MinMaxScaler normalization.
```
Comment (lmmentel) [inline sktime/transformations/series/dobin.py:32]:
```suggestion
    ``MinMaxScaler`` normalization.
```

File: sktime/transformations/series/dobin.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+    Parameters
+    ----------
+    frac : int, float (default=0.95)
```
Comment (lmmentel) [inline sktime/transformations/series/dobin.py:36]: Since it's a quantile maybe leave `int` out and keep `float` only

File: sktime/transformations/series/dobin.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        The cut-off quantile for Y space
+    k : int (default=None)
+        Number of nearest neighbours considered, with a default value None calculated
+        as 5% of the number of observations with a cap of 20.
```
Comment (lmmentel) [inline sktime/transformations/series/dobin.py:40]: Any chance you could reformulate what the default value is, it's a bit confusing.

File: sktime/transformations/series/dobin.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+        basis.columns = ["".join(["DB", str(i)]) for i in range(len(basis.columns))]
+        self._basis = basis
+        coords.columns = ["".join(["DB", str(i)]) for i in range(len(coords.columns))]
+        self._coords = coords
+
+        return self
+
+    def _transform(self, X, y=None):
+
```
Comment (lmmentel) [inline sktime/transformations/series/dobin.py:164]: missing docstring

File: sktime/transformations/series/dobin.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        self,
+        frac=0.95,
+        k=None,
+    ):
+        self.frac = frac
+        self.k = k
+        super(DOBIN, self).__init__()
+
+    def _fit(self, X, y=None):
+
```
Comment (lmmentel) [inline sktime/transformations/series/dobin.py:92]: missing docstring

File: sktime/transformations/series/dobin.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+                "Refitting with new input data, not storing updated public class "
+                "attributes. For this, explicitly use fit(X) or fit_transform(X)."
+            )
+            return new_dobin._coords
+
+        return self._coords
+
+
+def close_distance_matrix(X, k, frac):
+    """Calculate distance between close pairs."""
```
Comment (lmmentel) [inline sktime/transformations/series/dobin.py:184]: Could you expand the docstring to cover argument types and descriptions and the output?

**Review by lmmentel — CHANGES_REQUESTED** [review]

**Comment by KatieBuc** [comment]
@fkiraly ready to merge?

**Review by fkiraly — CHANGES_REQUESTED** [review]
In-principle happy with this.

My only change request is to ensure the docstring references the algorithm precisely.
Can you describe clearly what thet output of `transform` is, in the class docstring?

Also, what are "basistors"? Is that at typo or a term from the paper, if the latter it needs to be defined.

**Review by lmmentel — APPROVED** [review]

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!

In the class doc-string, could you please add:
* a reference to precisely which algorithm is implemented (algorithm X from page Y)
* an explanation of how the algorithm differs from the reference, e.g., normalization
* optimally: how parameters map onto symbols in the paper
* how the outliers are encoded as output of `transform`

**Comment by KatieBuc** [comment]
Thanks @fkiraly - this should address everything now

**Review by fkiraly — APPROVED** [review]
All great! Sorry for not merging, was preparing the release.
